lib/recursive-methods.rb

```diff
@@ -1,49 +1,91 @@
 # Authoring recursive algorithms. Add comments including time and space complexity for each method.
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n) where n is the value of the input
+# Space complexity: O(1) because there's nothing storing the result
 def factorial(n)
-    raise NotImplementedError, "Method not implemented"
+  return 1 if n == 0
+  return 1 if n == 1
+  raise ArgumentError if n < 0
+  return n * factorial(n - 1)
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n) where n is the length of s
+# Space complexity: O(n) where n is the length of s
 def reverse(s)
-    raise NotImplementedError, "Method not implemented"
+  return s if s.length == 0
+  last = s[-1]
+  s = s[0..-2]
+  reversed_string = last + reverse(s)
+  return reversed_string
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n) where n is the length of the string
+# Space complexity: O(1)
 def reverse_inplace(s)
-    raise NotImplementedError, "Method not implemented"
+  return s if s.length == 0
+  return s if s.length == 1
+
+  reverse_helper(s, 0, s.length - 1)
+  return s
 end
 
-# Time complexity: ?
-# Space complexity: ?
+def reverse_helper(s, first_index, last_index)
+  return if first_index > last_index
+
+  temp = s[first_index]
+  s[first_index] = s[last_index]
+  s[last_index] = temp
+  reverse_helper(s, first_index + 1, last_index - 1)
+end
+
+# Time complexity: O(n) where n is the value of the input
+# Space complexity: O(1)
 def bunny(n)
-    raise NotImplementedError, "Method not implemented"
+  return 0 if n == 0
+  return 2 if n == 1
+  return 2 + bunny(n - 1)
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n) where n is the length of s
+# Space complexity: O(1)
 def nested(s)
-    raise NotImplementedError, "Method not implemented"
+  return true if s == ""
+  return false if !s.include?("(")
+  return false if !s.include?(")")
+  s.sub!("(", "")
+  s.sub!(")", "")
+  return nested(s)
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n) where n is the length of the array
+# Space complexity: O(1) because only the variable last storing one element of the array
 def search(array, value)
-    raise NotImplementedError, "Method not implemented"
+  return false if array.empty?
+  return true if array[0] == value
+  last = array.pop
+  return true if last == value
+  return search(array, value)
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n) where is the length of s
+# Space complexity: O(1) because there's no variable storing the string
 def is_palindrome(s)
-    raise NotImplementedError, "Method not implemented"
+  return true if s.empty?
+  if s[0] == s[-1]
+    s.sub!(s[0], "")
```

Comment:
`sub` is an O(n) algorithm, so this is an O(n) operation you are doing n times so the full algorithm is O(n^2) for both time and space complexity. 

Instead think about tracking the left and right indexes adjusting them each recursive call.